Mistral-7B WH demo

mtairum opened the pull request:
Part of the Mistral-7B and Mixtral bringup from issue #5337 .

This includes a working demo of Mistral-7B, based on the original source-code from Mistral AI. The model was ported to TTNN.

We include tests for each part of the model: RMS Layernorm, MLP, Attention, single decoder block and full model.
We also include a full demo code that generates questions for 32 users based on prompt file passed by the user (and included within the demo files).

This implementation supersedes the existent ones present in `models/experimental/mistral` (ttlib) and `models/experimental/functional_mistral` (ttnn).

Additionally, we also include a small perf report tool under `models/perf/perf_report.py`. This takes perf .csv files and generates a cleaner report. Let me know if this should be included in a separate PR.

mtairum commented:
Regarding checkpoint files, how do we integrate the Mistral-7B tests on CI? 
They can be downloaded directly from Mistral website, as mentioned in our README provided with the model.
 
We're relying on `mistral7b/tt/model_config.py` to pick the right consolidated weights and tokenizer (general or instruct), so that will have to be updated with the correct default location for CI.

mtairum commented:
Regarding documentation, should we add a link to the Mistral demo to https://tenstorrent-metal.github.io/tt-metal/latest/get_started/get_started.html#explore-the-falcon-7b-demo ?

TT-BrianLiu commented on models/demos/mistral7b/tests/test_mistral_model.py:
```diff
@@ -0,0 +1,238 @@
+                        # if not does_pass:
+                        # all_tests_pass = False
+
+        print("[ttnn generation User 0] ", "".join(tokenizer.decode(all_outputs)))
```
Remove prints. If you want to keep these around as debug, use `logger.trace`

TT-BrianLiu commented on models/demos/mistral7b/tt/mistral_attention.py:
```diff
@@ -0,0 +1,348 @@
+                    self.max_batch_size,
+                    self.n_kv_heads // self.num_devices,
+                    # self.sliding_window,
+                    512,  # TODO Update the initial cache size when scaling up (Target = window size == 4096)
```
Is 512 the seq_len? Should this be something we pass in to the model?

mtairum replied:
Yes. This should be equal to `self.sliding_window` == 4096, but that would be too large to fit on DRAM for a single chip.

I could add this as a parameter to model_config instead, to keep the attn module cleaner.

mtairum replied:
I moved this to model_config.

TT-BrianLiu reviewed (commented):
Left some minor comments. Also, I think it would be better if memory configs, program configs, etc... is refactored into `model_config.py`. It will make it much easier to play around with different setups for multi-chip.

mtairum commented on models/demos/mistral7b/tt/mistral_model.py:
```diff
@@ -0,0 +1,81 @@
+            x = layer(x, start_pos, current_pos, attn_masks, rot_mat)
+
+        x = self.norm(x)
+        output = ttnn.linear(x, self.output_weight, core_grid=ttnn.CoreGrid(x=8, y=8))
```
Should we specify instead the grid_size in model_config to something like:

 max_grid_size = (8, 8) if device.arch() == ttl.device.Arch.WORMHOLE_B0 else (9, 12)

TT-BrianLiu replied:
does this op have to run on max 8, 8 grid? If that's the case, it won't work on any harvested wormhole

mtairum replied:
No. I can reduce to (x=4,y=8) grid. X==rows, right?

TT-BrianLiu replied:
x is along the row, so it's the number of columns. `(x=4, y=8)` should fail on harvested wormhole. Unless you are using these coords to explicitly set sharding specs, we should use default device compute grid instead of hardcoding like this.

Please change.

mtairum replied:
Pushed the change to pick device's grid size instead.

```
grid_size = device.compute_with_storage_grid_size()
self.max_grid_size = ttnn.CoreGrid(x=grid_size.x, y=grid_size.y)
```

TT-BrianLiu replied:
this still works for you? If so, good to merge.

FYI, ttnn coordinates are `(y, x)`, so if you want to avoid confusion you should probably flip the coordinates for ttnn. Constructing the coordinates with kwargs is fine as well, but just know that it's actually `(y, x)` by default.

eyonland reviewed (changes requested):
Looks amazing.  Can we get run_performance.sh updated to collect both the e2e and device times?

mtairum commented:
Refactored the model configs in the latest commit.

mtairum commented:
@eyonland I've added `models/demos/mistral7b/tests/test_mistral_perf.py` and added mistral-7B (WH only) on `run_performance.sh` script (Haven't tested on grayskull).

Can you please review the latest changes and let me know what you think?

tt-rkim commented on models/perf/perf_report.py:
```diff
@@ -0,0 +1,129 @@
+import argparse
+
+
+def analyze_perf_csv(csv_file, layers, show_all):
```
What is the point of this file? It's not used - is it an internal tool? Isn't this the point of the perf tests? If there's no need for this and people can use the perf tests, then please delete this.

mtairum replied:
I've removed this file.

tt-rkim commented:
@mtairum I'll approve once you post a passing models performance pipeline that runs your test.

tt-rkim commented:
Please run the performance pipeline and post a passing run here.

tt-rkim commented:
@mtairum @yieldthought 
There is a conflict on this branch with some perf pipeline optimizations for your models team.

Please let me know if you would like me to rebase for you to resolve.

tt-rkim reviewed (approved):
Please update the bash script to gate the mistral call for grayskull and update CI perf time

yieldthought commented:
> Looks amazing. Can we get run_performance.sh updated to collect both the e2e and device times?

Added under `run_device_perf_models()` and `run_perf_models_llm_javelin()`

mtairum committed:
#5337: Add demo mistral-7b in ttnn

mtairum committed:
#5337: Update mistral perf time
